test/suggestion.cpp

```diff
@@ -358,8 +358,36 @@ namespace {
     // We should get only one result
     std::vector<std::string> expectedResult = {
                                                 "Article Target",
-                                                "Article Redirect 1",
-                                                "Article Redirect 2"
+                                              };
+    ASSERT_EQ(resultSet, expectedResult);
+  }
+
+  TEST(Suggestion, checkRedirectionChain) {
+    /*
+     * As of now, we do not handle redirection chain. So if we have articles such
+     * as A->B->C. Even if A B and C are essentially the same articles, They won't
+     * get collapsed as one.
+     */
+    TempZimArchive tza("testZim");
+    zim::writer::Creator creator;
+    creator.configIndexing(true, "en");
+    creator.startZimCreation(tza.getPath());
+
+    auto item = std::make_shared<TestItem>("testPath", "text/html", "Article Target");
+    creator.addItem(item);
+    creator.addRedirection("redirectionPath1", "Article Redirect 1", "testPath");
+    creator.addRedirection("redirectionPath2", "Article Redirect 2", "redirectionPath1");
+
+    creator.addMetadata("Title", "Test zim");
+    creator.finishZimCreation();
+
+    zim::Archive archive(tza.getPath());
+    std::vector<std::string> resultSet = getSuggestions(archive, "Article", archive.getEntryCount());
+
+    // We should get only one result
+    std::vector<std::string> expectedResult = {
+                                                "Article Target",
+                                                "Article Redirect 1"
```

Comment:
This isn't all. When I change the query to "Article redirect" the search result is the same as above and includes "Article Target".

Reply:
The test is working as expected. I believe this apparent "issue" is caused by the "resolve redirect" step. In the current version, 
- when I make a query `"Article Redirect 1"`, the result set I get is `{"Article Target"}`
- when I make a query `"Article Redirect 3"`, the result set I get is `{"Article Redirect 2"}`. 

It seems like when the `setRedirect()` method is called during "resolve redirects", the target index completely takes over the redirect index while maintaining the terms of the redirect in the document(ie whenever the redirect item is requested, the target's index is used). That's why we are getting these results. So the result set I am getting in the unit test is essentially:
```
"Article Target" (originally: article target)
"Article Redirect 1" (originally: article redirect 2, but index is of its target "article redirect 1")
"Article Redirect 2" (originally: article redirect 3, but index is of its target "article redirect 2")
``` 
Since this is a question of what behavior is to be expected, I need your suggestions @mgautierfr @veloman-yunkan @kelson42 .

Reply:
It is not the behavior expected (at least not by me).

In you test we are indexing three things :
- `Article Target` with path `testPath` (the item)
- `Article Redirect 1` with path `testPath` (the redirection `redirectionPath1`)
- `Article Redirect 2` with path `redirectionPath1` (the redirection `redirectionPath2`)

When we search for "Article" we should have (internally) :
- A "group" for `testPath` [`Article Target`, `Article Redirect 1`]
- A "group" for `redirectionPath1` [`Article Redirect 2`]

As we keep only one element per group we should have the result (`Article Target`, `Article Redirect 2`) or (`Article Redirect 1`, `Article Redirect 2`), depending of the score of `Article Target` and `Article Redirect 1`.

If we search for "Redirect" we should have (internally) :
- A "group" for `testPath` [`Article Redirect 1`]
- A "group" for `redirectionPath1` [`Article Redirect 2`]

As we keep only one element per group we should have the result (`Article Redirect 1`, `Article Redirect 2`).

But we should never have a result (`Article Target`, `Article Redirect 1`) as they are sharing the same key(path).

Reply:
Understood, I have changed the value to store the path/redirectPath and the data in the index is storing the original path. Now we are getting the expected behavior.